decompiler/pipeline/controlflowanalysis/restructuring_commons/condition_aware_refinement_commons/switch_extractor.py

```diff
@@ -4,54 +4,46 @@
     BaseClassConditionAwareRefinement,
 )
 from decompiler.pipeline.controlflowanalysis.restructuring_options import RestructuringOptions
-from decompiler.structures.ast.ast_nodes import ConditionNode, FalseNode, SeqNode, TrueNode
+from decompiler.structures.ast.ast_nodes import ConditionNode, FalseNode, SeqNode, SwitchNode, TrueNode
 from decompiler.structures.ast.syntaxforest import AbstractSyntaxForest
+from decompiler.structures.logic.logic_condition import LogicCondition
 
 
 class SwitchExtractor(BaseClassConditionAwareRefinement):
     """Extract switch nodes from condition nodes if the condition node is irrelevant for the switch node."""
 
-    def __init__(self, asforest: AbstractSyntaxForest, options: RestructuringOptions):
-        """
-        self.current_cond_node: The condition node which we consider to extract switch nodes.
-        """
-        super().__init__(asforest, options)
-        self._current_cond_node: Optional[ConditionNode] = None
-
     @classmethod
     def extract(cls, asforest: AbstractSyntaxForest, options: RestructuringOptions):
-        """
-        Extract switch nodes from condition nodes, i.e., if a switch node is a branch of a condition node whose condition is redundant for
-        the switch node, we extract it from the condition node.
-        """
+        """Extract switch nodes from condition nodes, or sequence-nodes with a non-trivial reaching-condition."""
         switch_extractor = cls(asforest, options)
-        for condition_node in asforest.get_condition_nodes_post_order(asforest.current_root):
-            switch_extractor._current_cond_node = condition_node
-            switch_extractor._extract_switches_from_condition()
+        for switch_node in list(asforest.get_switch_nodes_post_order(asforest.current_root)):
+            while switch_extractor._switch_can_be_extracted(switch_node):
+                pass
 
-    def _extract_switches_from_condition(self) -> None:
-        """Extract switch nodes in the true and false branch of the given condition node."""
-        if self._current_cond_node.false_branch:
-            self._try_to_extract_switch_from_branch(self._current_cond_node.false_branch)
-        if self._current_cond_node.true_branch:
-            self._try_to_extract_switch_from_branch(self._current_cond_node.true_branch)
-        if self._current_cond_node in self.asforest:
-            self._current_cond_node.clean()
-
-    def _try_to_extract_switch_from_branch(self, branch: Union[TrueNode, FalseNode]) -> None:
+    def _switch_can_be_extracted(self, switch_node: SwitchNode) -> bool:
```

Comment:
I think the function name is a bit misleading, indicating that it is only checked if the extraction is possible or not. Instead, the extraction is actually performed.

Reply:
renamed

Reply:
Well, I still don't particularly like the naming, but sure :D